airflow/sentry.py

```diff
@@ -0,0 +1,203 @@
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Sentry Integration"""
+
+
+from typing import Any
+from functools import wraps
+
+from airflow import configuration as conf
+from airflow.utils.db import provide_session
+from airflow.utils.log.logging_mixin import LoggingMixin
+from airflow.utils.state import State
+
+
+log = LoggingMixin().log
+
+
+class DummySentry:
+    """
+    Blank class for Sentry.
+    """
+
+    @classmethod
+    def add_tagging(cls, task_instance):
+        """
+        Blank function for tagging.
+        """
+
+    @classmethod
+    def add_breadcrumbs(cls, session=None):
+        """
+        Blank function for breadcrumbs.
+        """
+
+    @classmethod
+    def format_run_task(cls, run):
+        """
+        Blank function for formatting a TaskInstance._run_raw_task.
+        """
+        return run
+
+
+@provide_session
+def get_task_instances(dag_id, task_ids, execution_date, session=None):
+    """
+    Retrieves task instance based on dag_id and execution_date.
+    """
+    from airflow.models.taskinstance import TaskInstance  # Avoid circular import
+
+    if session is None or not task_ids:
+        return []
+
+    TI = TaskInstance
+    ti = (
+        session.query(TI)
+        .filter(
+            TI.dag_id == dag_id,
+            TI.task_id.in_(task_ids),
+            TI.execution_date == execution_date,
+        )
+        .all()
+    )
+    return ti
+
+
+class ConfiguredSentry:
+    """
+    Configure Sentry SDK.
+    """
+
+    SCOPE_TAGS = frozenset(("task_id", "dag_id", "execution_date", "operator"))
+    SCOPE_CRUMBS = frozenset(
+        ("dag_id", "task_id", "execution_date", "state", "operator", "duration")
+    )
+
+    def __init__(self):
+        """
+        Initialize the Sentry SDK.
+        """
+        ignore_logger("airflow.task")
+        ignore_logger("airflow.jobs.backfill_job.BackfillJob")
+        executor_name = conf.get("core", "EXECUTOR")
+
+        sentry_flask = FlaskIntegration()
+
+        # LoggingIntegration is set by default.
+        integrations = [sentry_flask]
+
+        if executor_name == "CeleryExecutor":
+            from sentry_sdk.integrations.celery import CeleryIntegration
+            from sentry_sdk.integrations.tornado import TornadoIntegration
+
+            sentry_celery = CeleryIntegration()
+            sentry_tornado = TornadoIntegration()
+            integrations += [sentry_celery, sentry_tornado]
+
+        self.task = None
```

Comment:
Do we need to store these attributes on the instance? Maybe I'm missing something, but it seems like it would be simpler to pass `task_instance` to both `add_tagging` and `add_breadcrumbs` and not keep state on the instance at all.

Reply:
Sounds good